In [1]:
class Atbash:
    alphabet = "абвгдеёжзийклмнопрстуфхцчшщъыьэюя"

    def __init__(self):
        lowercase_code = {x: y for x, y in zip(self.alphabet, self.alphabet[::-1])}
        uppercase_code = {x.upper(): y.upper() for x, y in zip(self.alphabet, self.alphabet[::-1])}
        self._encode = lowercase_code
        self._encode.update(uppercase_code)

    def encode(self, text):
        return ''.join([self._encode.get(char, char) for char in text])


cipher = Atbash()
line = input()
while line != '.':
    print(cipher.encode(line))
    line = input()

   Жцко Иъчяоа¶ Жцко Иъчяоа — вмр эцы жцкоя прынмясрэфц, э фрмрорт фяшыдх нцтэру э рмфодмрт мъфнмъ чятъсаъмна нцтэрурт, сяйрыаёцтна ся съфрмрорт прнмрассрт зцнуъ прчцицх уъэъъ цуц пояэъъ съьр э яукяэцмъ. Сяпоцтъо, э жцкоъ нр ныэцьрт эпояэр ся 3, Я юдуя юд чятъсъся ся Ь, Ю нмясъм Ы, ц мяф ыяуъъ.  Жцко сячэяс э зънмг оцтнфрьр цтпъоямроя Ьяа Буца Иъчяоа, цнпругчрэяэжъьр ъьр ыуа нъфоъмсрх пъоъпцнфц нр нэрцтц ьъсъояуятц.  Нуъылбёяа зянмг ояюрмд чяжцкорэяся жцкорт Иъчяоа. Ырпцжцмъ тъмры decode ц оянжцколхмъ нуъылбёцх оячыъу уяюроямросрх ояюрмд. Прылтяхмъ, прзътл эят съ нррюёцуц фубз жцкорэясца ц змр эят н вмцт ыъуямг.


  Шифр Цезаря¶ Шифр Цезаря — это вид шифра подстановки, в котором каждый символ в открытом тексте заменяется символом, находящимся на некотором постоянном числе позиций левее или правее него в алфавите. Например, в шифре со сдвигом вправо на 3, А была бы заменена на Г, Б станет Д, и так далее.  Шифр назван в честь римского императора Гая Юлия Цезаря, использовавшего его для секретной переписки со своими генералами.  Следующая часть работы зашифрована шифром Цезаря. Допишите метод decode и расшифруйте следующий раздел лабораторной работы. Подумайте, почему вам не сообщили ключ шифрования и что вам с этим делать.


KeyboardInterrupt: Interrupted by user

Шифр Цезаря¶ 
  
  Шифр Цезаря — это вид шифра подстановки, в котором каждый символ в открытом тексте заменяется символом, находящимся на некотором постоянном числе позиций левее или правее него в алфавите. Например, в шифре со сдвигом вправо на 3, А была бы заменена на Г, Б станет Д, и так далее.  Шифр назван в честь римского императора Гая Юлия Цезаря, использовавшего его для секретной переписки со своими генералами.  Следующая часть работы зашифрована шифром Цезаря. Допишите метод decode и расшифруйте следующий раздел лабораторной работы. 
  
  Подумайте, почему вам не сообщили ключ шифрования и что вам с этим делать.

In [10]:
class Caesar:
    alphabet = "яюэьыъщшчцхфутсрпонмлкйизжёедгвба"

    def __init__(self, key):
        self._encode = dict()
        self._decode = dict()
        for i in range(len(self.alphabet)):
            letter = self.alphabet[i]
            encoded = self.alphabet[(i + key) % len(self.alphabet)]
            self._encode[letter] = encoded
            self._encode[letter.upper()] = encoded.upper()
            self._decode[encoded] = letter
            self._decode[encoded.upper()] = letter.upper()

    def encode(self, text):
        return ''.join([self._encode.get(char, char) for char in text])

    def decode(self, line):
        return ''.join([self._decode.get(char, char) for char in line])


key = int(input('Введите ключ:'))
cipher = Caesar(key)
line = input()
while line:
    print(cipher.encode(line))
    line = input()

Введите ключ: 19
 Кыжг вгбдебь ътячан Вбъцгтфюсчя д гтдкыжгбфэбь гтъцчют!  Ыетэ, фн цбхтцтюыдо вбйчяё кыжг Ичътгс ач сфюсчедс эгывебдебьэыя: дюыкэбя ятют яблабдео яабщчдефт эюрйчь ы аёщань эюрй ючхэб атьеы ячебцбя вбюабхб вчгчубгт.  Ябщаб юы ёфчюыйыео эгывебдебьэбдео, ач ячасс ячебц кыжгбфтаыс? Цт, ябщаб. Чдюы ътячасео бцыа дыяфбю тюжтфыет ат бвгчцчюшаань цгёхбь дыяфбю ебхб щч тюжтфыет вб этэбь-еб етуюыич ътяча, еб дтят етуюыит ътяча ы сфюсчедс эюрйбя.  |т|у|ф|х|ц|ч|ш|щ|ъ|ы|ь|э|ю|я|а|б|в|г|д|е|ё|ж|з|и|й|к|л|м|н|о|п|р|с|  |о|ъ|и|ф|э|у|л|з|ж|в|г|с|ю|п|т|ч|н|к|ё|я|м|щ|ш|б|х|ь|а|ц|ы|е|й|р|д|  Яабщчдефб эюрйчь — пеб яабщчдефб фбъябщанз етуюыи вгбденз ътяча. Цюс гёддэбхб тюжтфыет яблабдео яабщчдефт етуюыи вгбденз ътяча гтфат жтэебгытюё бе 33. 33! = 8683317618811886495518194401280000000  Чдюы егтеыео ат вгбфчгэё бцабхб фтгытает 0.000001 дчэёацн, вбюёйыедс 2.8e+23 юче. Ябщче вбэтътеодс, йеб кыжг вгбдебь ътячан фвбюач эгывебдебьэыь, бцатэб пеб ач етэ.  Чхб цбдетебйаб вгбдеб фъюбятео вгы вбяблы 

Шифр простой замены Поздравляем с расшифровкой раздела!  Итак, вы догадались почему шифр Цезаря не является криптостойким: слишком мала мощность множества ключей и нужный ключ легко найти методом полного перебора.  Можно ли увеличить криптостойкость, не меняя метод шифрования? Да, можно. Если заменять один символ алфавита на определённый другой символ того же алфавита по какой-то таблице замен, то сама таблица замен и является ключом.  |а|б|в|г|д|е|ё|ж|з|и|й|к|л|м|н|о|п|р|с|т|у|ф|х|ц|ч|ш|щ|ъ|ы|ь|э|ю|я|  |ь|з|ц|в|к|б|щ|х|ф|п|р|я|л|э|а|е|ы|ш|у|м|ъ|ж|ё|о|г|й|н|д|и|т|ч|ю|с|  Множество ключей — это множество возможных таблиц простых замен. Для русского алфавита мощность множества таблиц простых замен равна факториалу от 33. 33! = 8683317618811886495518194401280000000  Если тратить на проверку одного варианта 0.000001 секунды, получится 2.8e+23 лет. Может показаться, что шифр простой замены вполне криптостойкий, однако это не так.  Его достаточно просто взломать при помощи частотного анализа

KeyboardInterrupt: Interrupted by user

Шифр простой замены 

Поздравляем с расшифровкой раздела!  

Итак, вы догадались почему шифр Цезаря не является криптостойким: слишком мала мощность множества ключей и нужный ключ легко найти методом полного перебора.  

Можно ли увеличить криптостойкость, не меняя метод шифрования? Да, можно. Если заменять один символ алфавита на определённый другой символ того же алфавита по какой-то таблице замен, то сама таблица замен и является ключом.  

|а|б|в|г|д|е|ё|ж|з|и|й|к|л|м|н|о|п|р|с|т|у|ф|х|ц|ч|ш|щ|ъ|ы|ь|э|ю|я|  

|ь|з|ц|в|к|б|щ|х|ф|п|р|я|л|э|а|е|ы|ш|у|м|ъ|ж|ё|о|г|й|н|д|и|т|ч|ю|с|  

Множество ключей — это множество возможных таблиц простых замен. Для русского алфавита мощность множества таблиц простых замен равна факториалу от 33. 33! = 8683317618811886495518194401280000000  

Если тратить на проверку одного варианта 0.000001 секунды, получится 2.8e+23 лет. Может показаться, что шифр простой замены вполне криптостойкий, однако это не так.  Его достаточно просто взломать при помощи частотного анализа. Дело в том, что частота появления заданной буквы алфавита в достаточно длинных текстах одна и та же для разных текстов одного языка. Если в шифротексте будет символ с вероятностью появления, аналогичной стандартной для языка, то можно предположить, что он и является указанной зашифрованной буквой.  

Метод частотного криптоанализа известен с IX-го века, хотя наиболее известным случаем его применения в реальной жизни, возможно, является дешифровка египетских иероглифов в 1822 году.  

Итак, следующая часть работы зашифрована при помощи следующей программы:

Программу для частотного анализа следует написать самостоятельно. Успехов!

In [4]:
text = input()
alpha = "абвгдеёжзийклмнопрстуфхцчшщъыьэюя"
freq = "оеаинтсрвлкмдпуяыьгзбчйхжшюцщэфъё"

count = {ch: 0 for ch in alpha}
for i in text.lower():
    if i in count:
        count[i] += 1

sorted_freq = sorted(alpha, key=lambda ch: -count[ch])

decode_map = {}
for i in range(33):
    decode_map[sorted_freq[i]] = freq[i]
    decode_map[sorted_freq[i].upper()] = freq[i].upper()


print(decode_map)

 ыыы


{'ы': 'о', 'Ы': 'О', 'а': 'е', 'А': 'Е', 'б': 'а', 'Б': 'А', 'в': 'и', 'В': 'И', 'г': 'н', 'Г': 'Н', 'д': 'т', 'Д': 'Т', 'е': 'с', 'Е': 'С', 'ё': 'р', 'Ё': 'Р', 'ж': 'в', 'Ж': 'В', 'з': 'л', 'З': 'Л', 'и': 'к', 'И': 'К', 'й': 'м', 'Й': 'М', 'к': 'д', 'К': 'Д', 'л': 'п', 'Л': 'П', 'м': 'у', 'М': 'У', 'н': 'я', 'Н': 'Я', 'о': 'ы', 'О': 'Ы', 'п': 'ь', 'П': 'Ь', 'р': 'г', 'Р': 'Г', 'с': 'з', 'С': 'З', 'т': 'б', 'Т': 'Б', 'у': 'ч', 'У': 'Ч', 'ф': 'й', 'Ф': 'Й', 'х': 'х', 'Х': 'Х', 'ц': 'ж', 'Ц': 'Ж', 'ч': 'ш', 'Ч': 'Ш', 'ш': 'ю', 'Ш': 'Ю', 'щ': 'ц', 'Щ': 'Ц', 'ъ': 'щ', 'Ъ': 'Щ', 'ь': 'э', 'Ь': 'Э', 'э': 'ф', 'Э': 'Ф', 'ю': 'ъ', 'Ю': 'Ъ', 'я': 'ё', 'Я': 'Ё'}


In [11]:
class Monoalphabet:
    alphabet = "егпжонщюэбэчмяхыцилдшарсйзвьёктф"

    def __init__(self, keytable):
        lowercase_code = {x: y for x, y in zip(self.alphabet, keytable)}
        uppercase_code = {x.upper(): y.upper() for x, y in zip(self.alphabet, keytable)}
        self._encode = lowercase_code
        self._encode.update(uppercase_code)
        self._decode = {}
        for plain, cipher in zip(self.alphabet, keytable):
            self._decode[cipher] = plain
            self._decode[cipher.upper()] = plain.upper()

    def encode(self, text):
        return ''.join([self._encode.get(char, char) for char in text])

    def decode(self, line):
        return ''.join([self._decode.get(char, char) for char in line])


keytable = "шифрвженатакдльйоячёспгмзыхбцэую"

cipher = Monoalphabet(keytable)
line = input()
while line:
    print(cipher.encode(line))
    line = input()

 Егпж Огнщющжэ (ацягэяпэогбюцы йэсщюз) Мэяхющыегс ажцмцянщюгщс егпжцо ажцшбцы йэсщюз иояищбши сюцрцэяпэогбюзщ егпжз. Эьт Эях-Чгюмг о шоцгв жэьцбэв ацчэйэя, лбц цьзлюзщ сцюцэяпэогбюзщ егпжз мцоцяхюц-бэчг ажцшбц ацммэфбши лэшбцбюцст чжгабцэюэягйт г ащжозс ажщмяцнгя гшацяхйцоэбх сюцрцэяпэогбюзщ егпжз. О Щожцащ бэчгщ егпжз ьзяг оащжозщ цагшэюз о 1467 рцмт гбэяхиюшчгс эжвгбщчбцжцс Ящцю Ьэббгшбэ Эяхьщжбг. О XVI ощчщ ющсщёчгы эььэб Гцрэюю Бжгбщсгы о шоцщы чюгрщ “Шбщюцржэпги” ажщмшбэогя швщст ацягэяпэогбюцрц егпжцоэюги о огмщ бэьягёз. Ьцящщ шяцнюзы оэжгэюб ш гшацяхйцоэюгщс шсщеэююзв эяпэогбцо ьзя цагшэю о 1563 рцмт Мнэсьэббгшбэ мщяяэ Ацжбэ о щрц чюгрщ “Ажц шчжзбтф йюэлгсцшбх цбмщяхюзв ьтчо”. Ацшящмюгс шяцоцс о жэйогбгг ацягэяпэогбюзв егпжцо сцнюц шлгбэбх жцбцжюзщ сэегюз, ажгсщжцс чцбцжцы сцнюц шлгбэбх ющсщёчтф сэегют Enigma, жэйжэьцбэююэи о 1917 р. Штбх ацягэяпэогбюзв егпжцо йэчяфлщюэ о сюцрцчжэбюцс ажгсщющюгг жэйяглюзв егпжцо ажцшбцы йэсщюз ч цажщмщящююцст лгшят ьтчо егпжтщсцрц бщчшбэ. Бц щшб

Шифр Виженера (полиалфавитной замены) Дальнейшим продолжением шифров простой замены является многоалфавитные шифры. Абу Аль-Кинди в своих работах показал, что обычные моноалфавитные шифры довольно-таки просто поддаются частотному криптоанализу и первым предложил использовать многоалфавитные шифры. В Европе такие шифры были впервые описаны в 1467 году итальянским архитектором Леон Баттиста Альберти. В XVI веке немецкий аббат Иоганн Тритемий в своей книге “Стенография” представил схему полиалфавитного шифрования в виде таблицы. Более сложный вариант с использованием смешанных алфавитов был описан в 1563 году Джамбаттиста делла Порта в его книге “Про скрытую значимость отдельных букв”. Последним словом в развитии полиалфавитных шифров можно считать роторные машины, примером которой можно считать немецкую машину Enigma, разработанная в 1917 г. Суть полиалфавитных шифров заключена в многократном применении различных шифров простой замены к определенному числу букв шифруемого текста. То есть

KeyboardInterrupt: Interrupted by user

Шифр Виженера (полиалфавитной замены) Дальнейшим продолжением шифров простой замены является многоалфавитные шифры. Абу Аль-Кинди в своих работах показал, что обычные моноалфавитные шифры довольно-таки просто поддаются частотному криптоанализу и первым предложил использовать многоалфавитные шифры. В Европе такие шифры были впервые описаны в 1467 году итальянским архитектором Леон Баттиста Альберти. В XVI веке немецкий аббат Иоганн Тритемий в своей книге “Стенография” представил схему полиалфавитного шифрования в виде таблицы. Более сложный вариант с использованием смешанных алфавитов был описан в 1563 году Джамбаттиста делла Порта в его книге “Про скрытую значимость отдельных букв”. Последним словом в развитии полиалфавитных шифров можно считать роторные машины, примером которой можно считать немецкую машину Enigma, разработанная в 1917 г. Суть полиалфавитных шифров заключена в многократном применении различных шифров простой замены к определенному числу букв шифруемого текста. То есть к каждой букве по отдельности применяется один из шифров простой замены.  

Шифр Виженера состоит из последовательности нескольких шифров Цезаря с различными значениями сдвига. Для зашифровывания может использоваться таблица алфавитов, называемая квадрат (таблица) Виженера. Применительно к русскому алфавиту таблица Виженера составляется из строк по 33 символов, причём каждая следуюъая строка сдвигается на несколько позиций. Таким образом, в таблице получается 33 различных шифров Цезаря. На разных этапах кодировки шифр Виженера использует различные алфавиты из этой таблицы. На каждом этапе шифрования используются различные алфавиты, выбираемые в зависимости от символа ключевого слова. Например, если ключевое слово “САТ”, то первая буква открытого текста шифруется с использованием алфавита “С’, вторая “А”, третья “Т”, четвёртая снова “С” и так далее.  Программа шифрования шифром Виженера:    

Последний раздел работы зашифрован шифром Виженера с неизвестным кодовым словом. Подсказка для самых стойких криптоаналитиков: длина кодового слова 8. 

In [9]:
class Vigenere:
    alphabet = "абвгдеёжзийклмнопрстуфхцчшщъыьэюя"

    def __init__(self, keyword):
        self.alphaindex = {ch: index for index, ch in enumerate(self.alphabet)}
        self.key = [self.alphaindex[letter] for letter in keyword.lower()]

    def caser(self, letter, shift):
        if letter in self.alphaindex:
            index = (self.alphaindex[letter] + shift) % len(self.alphabet)
            cipherletter = self.alphabet[index]
        elif letter.lower() in self.alphaindex:
            cipherletter = self.caesar(letter.lower(), shift).upper()
        else:
            cipherletter = letter
        return cipherletter

    def caesar(self, letter, shift):
        if letter in self.alphaindex:
            index = (self.alphaindex[letter] + shift) % len(self.alphabet)
            cipherletter = self.alphabet[index]
        elif letter.lower() in self.alphaindex:
            cipherletter = self.caesar(letter.lower(), shift).upper()
        else:
            cipherletter = letter
        return cipherletter

    def encode(self, line):
        ciphertext = []
        key_index = 0
        for letter in line:
            if letter.lower() in self.alphaindex:
                shift = self.key[key_index % len(self.key)]
                cipherletter = self.caesar(letter, shift)
                ciphertext.append(cipherletter)
                key_index += 1
            else:
                ciphertext.append(letter)
        return ''.join(ciphertext)

    def decode(self, line):
        plaintext = []
        key_index = 0
        for letter in line:
            if letter.lower() in self.alphaindex:
                shift = self.key[key_index % len(self.key)]
                # В ЭТОМ ШИФРЕ расшифровка = +shift
                plainletter = self.caesar(letter, -shift)
                plaintext.append(plainletter)
                key_index += 1
            else:
                plaintext.append(letter)
        return ''.join(plaintext)


keyword = 'мфти'
cipher = Vigenere(keyword)

line = input()
while line != '.':
    print(cipher.decode(line))
    line = input()

 йвылщф  ыцящгнюзрвхщдз щгхья дбахжтыи дгч эщтфхьтяхт дфыыачпг вчшэтфбффсявблы мыээф -- эчщдожящгцат ячрщюе еэжщм "саспбн". ёшщянъжн хсжбмыц ксбебкмвыз, хёвчшръчоффбхйдз о бтбхвтю йжбт ющгсх, эдшыаоратеъл шб ъхй вчэ.


энигма  интересующиеся могут почитать про реализацию сложного полиалфавитного шифра -- компьютерную модель шифра "энигмы". элементы методов шифрования, использовавшихся в машинах этой серии, используются до сих пор.


KeyboardInterrupt: Interrupted by user

Энигма

Интересующиеся могут почитать про реализацию сложного полиалфавитного шифра -- компьютерную модель шифра "энигмы". Элементы методов шифрования, использовавшихся в машинах этой серии, используются до сих пор.